psi4/driver/p4util/python_helpers.py

```diff
@@ -742,6 +743,148 @@ def anon(mol, role):
+        keys Python lists.
+
+    Examples
+    --------
+    >>> # [1] point charges on whole molecule (all equivalent)
+    >>> validate_external_potential([[0.5,0,0,1], [-0.5,0,0,-1]])
+    >>> validate_external_potential({"points": np.array([[0.5,0,0,1], [-0.5,0,0,-1]])})
+    >>> validate_external_potential({"C": {"points": psi4.core.Matrix.from_array(np.array([[0.5,0,0,1], [-0.5,0,0,-1]]))}})
+
+    >>> # [2] point charges on monoB and linker for FI-SAPT
+    >>> validate_external_potential({"b": [0.01,2,2,2], "c": [[0.5,0,0,1], [-0.5,0,0,-1]]})
+    >>> validate_external_potential({"B": {"points": np.array([0.01,2,2,2])}, "C": {"points": np.array([[0.5,0,0,1], [-0.5,0,0,-1]])}})
+
+    """
+    def validate_charge_list(lqxyz, diffuse: bool=False):
+        # check charge has form Q x y z, Q [x y z], Q x y z w, or Q [x y z] w
+
+        def validate_single_charge(chg):
+            try:
+                nchg = len(chg)
+            except TypeError:
+                return False
+
+            if (not diffuse) and nchg == 2:
+                return chg[0], chg[1][0], chg[1][1], chg[1][2]
+            elif diffuse and nchg == 3:
+                return chg[0], chg[1][0], chg[1][1], chg[1][2], chg[2]
+            elif (not diffuse) and nchg == 4:
+                return chg[0], chg[1], chg[2], chg[3]
+            elif diffuse and nchg == 5:
+                return chg[0], chg[1], chg[2], chg[3], chg[4]
+            else:
+                return False
+
+        flattened = [validate_single_charge(pt) for pt in lqxyz]
+
+        # reject if a charge line has wrong format or if spec is nested too deep or too shallow (can
+        #   happen due to flexible input format and backwards compatibility with points-only).
+        for itm in flattened:
+            if itm is False:
+                return False
+        try:
+            if np.array(flattened).ndim != 2:
+                return False
+        except ValueError:
+            return False
+        return flattened
+
+    def validate_potential_array(mat):
+        if isinstance(mat, core.Matrix):
+            npmat = mat.np
+        elif isinstance(mat, list):
+            npmat = np.array(mat)
+        elif isinstance(mat, np.ndarray):
+            npmat = mat
+        else:
+            return False
+
+        if npmat.ndim != 2 or npmat.shape[0] != npmat.shape[1]:
+            return False
+
+        return npmat.tolist()
+
+    frag_keys = set("ABC")
+    mode_keys_list = ["points", "diffuse", "matrix"]
+
+    if isinstance(external_potential, dict) and ({k.upper() for k in external_potential.keys()} <= frag_keys):
+        ep = {k.upper(): v for k, v in external_potential.items()}
+    else:
+        # assign "C" to mean whole molecule in the non-SAPT case
+        ep = {"C": external_potential}
+
+    # expand input structure: may be single points list or types list or types dict or fragments dict (for SAPT)
+    ep1 = {}
```

Comment:
The PR _seems_ fine, but I'm having a very hard time reading it.

Can you go through this file and use longer but more expressive variable names? Keeping track of `ep` vs `ep1` and `ep2` is a headache.

Reply:
sure, done